src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -332,115 +357,99 @@ def _post_process(
-        conf: Conf,
+        asu: AssemblyUtil,
+        ws: Workspace,
         upload_env_key: str,
         workspace_id: int,
         upload_dir: str,
         wait_to_upload_assemblies: dict[str, str],
+        batch_size: int,
+        output_dir: str,
 ) -> list[str]:
     """
     Upload assembly files to the target workspace in parallel using multiprocessing.
 
     Parameters:
-        conf: Conf object
+        asu: AssemblyUtil client
+        ws: Workspace client
         upload_env_key: environment variable key in uploaded.yaml file
         workspace_id: target workspace id
         upload_dir: a directory in collectionssource that creates new directories linking to sourcedata
         wait_to_upload_assemblies: a dictionary that maps assembly file name to assembly directory
+        batch_size: a number of files to upload per batch
+        output_dir: a directory in sourcedata/workspace to store new assembly entries
 
     Returns:
-        a list of assembly names that failed to upload
+        number of assembly files have been sucessfully uploaded from wait_to_upload_assemblies
     """
     assembly_files_len = len(wait_to_upload_assemblies)
-    print(f"Start uploading {assembly_files_len} assembly files with {conf.workers} workers\n")
+    print(f"Start uploading {assembly_files_len} assembly files\n")
 
-    # Put the assembly files in the input_queue
-    counter = 1
-    for assembly_name, host_assembly_dir in wait_to_upload_assemblies.items():
+    uploaded_count = 0
+    uploaded_fail = False
+    for assembly_tuples in _gen(wait_to_upload_assemblies, batch_size):
 
-        container_internal_assembly_path = os.path.join(
-            JOB_DIR_IN_ASSEMBLYUTIL_CONTAINER, assembly_name
-        )
-        conf.input_queue.put(
-            (
+        try:
+            batch_upas = _upload_assemblies_to_workspace(asu, workspace_id, assembly_tuples)
+            batch_uploaded_tuples = assembly_tuples
+        except Exception as e:
+            name_assemblyTuple_map = {
+                assembly_tuple.assembly_name: assembly_tuple for assembly_tuple in assembly_tuples
+            }
+            assembly_objects_in_ws = loader_helper.list_objects(
+                workspace_id,
+                ws,
+                loader_common_names.OBJECTS_NAME_ASSEMBLY,
+            )
+            print("assembly_objects_in_ws: ", assembly_objects_in_ws)
+            name_upa_map = {
+                object_info[1]: "{6}_{0}_{4}".format(*object_info) for object_info in assembly_objects_in_ws
+            }
+            batch_uploaded_objects_names = set(name_upa_map.keys()) & set(name_assemblyTuple_map.keys())
+
+            # update batch_uploaded_tuples and batch_upas
+            batch_upas = tuple([name_upa_map[name] for name in batch_uploaded_objects_names])
+            batch_uploaded_tuples = [name_assemblyTuple_map[name] for name in batch_uploaded_objects_names]
+            uploaded_fail = True
```

Comment:
There's another tricky thing here - the workspace may have previous uploads in it. If we're using the same names, but uploading a newer version of the assembly, this code will show the uploads succeeded when they didn't. 

What I would do is create a load_id (maybe a uuid) and put that in the object metadata. Then you can tell whether the object was already there or not.

Reply:
If we're using the same name and upload to the target workspace, it will fail based on the upload status from the yaml file.

Reply:
So how would things work if we wanted to upload, say, corrected assemblies with the same name?

Reply:
depending how many corrected assemblies to upload

1. Most objects in the target workspace
We can introduce a new --overwrite CLI. When specified, it basically ignores the upload status from the yaml file and still do the uploading. 

2. Few objects in the target workspace
We can introduce a new --batchfile CLI, a json/yaml file that specifies file_name: file_dir. If detected and not empty, we will upload only those valid ones.

Reply:
In the case where we're uploading over a previous upload and the upload fails, the --overwrite parameter would mean that the same genomes keep being uploaded until you get a run that has no failures. For the batchfile option, that means the user would have to edit the batch file again every time there was a failure

Reply:
> In the case where we're uploading over a previous upload and the upload fails

We discussed this already [here](https://github.com/kbase/collections/pull/426#discussion_r1427455965), when a specific object upload fails, user needs to check in and fix the failure. 

https://github.com/kbase/collections/pull/426#discussion_r1428249450 only addresses overwrite concern.

Reply:
The difference here is that the uploader can't use the yaml files to figure out where it left off

Reply:
Sorry, I went back to the codebase and found it will `skip` an object upload if having the same names, meaning corrected assemblies never get uploaded. `load_id` is a better approach than `overwrite` because it records upload version. Do we need `--batchfile` CLI? What is the size of corrected assemblies?

Reply:
If you have a load id you can tell what's part of this load and complete vs. part of a prior load. What a load id won't do is allow a user to specify a partial load. I'm not sure we have a use case for that though, so I think we can leave that for when we actually need it.

> What is the size of corrected assemblies?

Not sure what you mean here - the file size? Could theoretically be anything but likely similar to the prior version

Reply:
> What is the size of corrected assemblies?

I mean load size here. Yes, I was thinking a very small partial upload would need `--batchfile` logic.

Reply:
yeah, don't worry about that for now. I'm thinking more about total reloads for a new version of source data, for example

Reply:
So we will have the following file structure:

assembly_dir:
-- uploaded.yaml
-- metadata.json

metadata.json:
{
    load_id: [1, 2, 3]
}

Reply:
Do we need to keep the load ID in the yaml file? I was just thinking the loader would either take an arbitrary load id as a param or generate one and print it out. Whoever was doing the load would keep using that load id until it was done and then we wouldn't need it any more

Reply:
ah, where exactly do we put that load_id then?

Reply:
You don't. You provide it as a CLI parameter when you start the load and whenever you have to restart the load, then you throw it away when the load is complete.

At least that's the way I'm thinking, but maybe I'm missing something

Reply:
Updated!

Reply:
I've realized why I had qualms about only storing the load id in the yaml files. If there's anything that prevents the script from contacting the workspace (e.g. network partition, nginx / workspace / mongo down) then the yaml files can't be updated and so when the script is restarted it won't be able to pick up where it left off. 

That being said, we'd need to update assembly utils again to start storing a load ID with the object, and this change has already taken a really long time. So:

https://github.com/kbaseapps/AssemblyUtil/issues/93
https://github.com/kbase/collections/issues/588

Reply:
One way to reduce the impact of a failure would be to use smaller batch sizes so the yaml files are updated more regularly. I suspect this wouldn't have a huge impact on performance if they were still relatively large, like > 100.

Reply:
I personally don't like left unfinished work to other ppl.
> That being said, we'd need to update assembly utils again to start storing a load ID with the object

How much code you think we need to update in AssemblyUtil to add this load_id feature? (I could be wrong, but maybe not that much...)

Will any other repos require change as well?

Reply:
Just assembly utils, and we'd need to add the same feature to genome file utils. I suspect it wouldn't be that bad, just add `mapping<str, str> object_metadata;` to https://github.com/kbaseapps/AssemblyUtil/blob/develop/AssemblyUtil.spec#L180-L188 and then pass them to the save_objects calls

Reply:
fixed by https://github.com/kbaseapps/AssemblyUtil/pull/95 @Tianhao-Gu

Reply:
:+1: